src/lib/logical_query_plan/lqp_utils.hpp

```diff
@@ -5,24 +5,67 @@
 #include <queue>
 #include <set>
 #include <unordered_set>
+#include <utility>
+#include <vector>
 
 #include "logical_query_plan/abstract_lqp_node.hpp"
-#include "logical_query_plan/aggregate_node.hpp"
-#include "logical_query_plan/alias_node.hpp"
-#include "logical_query_plan/join_node.hpp"
-#include "logical_query_plan/limit_node.hpp"
-#include "logical_query_plan/predicate_node.hpp"
-#include "logical_query_plan/projection_node.hpp"
-#include "logical_query_plan/sort_node.hpp"
-#include "logical_query_plan/update_node.hpp"
 
 namespace opossum {
 
 class AbstractExpression;
+class AbstractLQPNode;
+class LQPSubqueryExpression;
+
 enum class LQPInputSide;
 
 using LQPMismatch = std::pair<std::shared_ptr<const AbstractLQPNode>, std::shared_ptr<const AbstractLQPNode>>;
 
+/**
+ * Data structure that maps LQPs to (multiple) subquery expressions that reference them.
+ *
+ * Purpose:
+ *  Mainly used by optimizer rules to optimize subquery LQPs more efficiently. In concrete, it helps
+ *  to optimize subquery LQPs ONLY ONCE, although being referenced by a list of subquery expressions.
+ *
+ * Why weak pointers for subquery expressions?
+ *  Referenced LQPs and subquery expressions might be objects to change after creating this data structure. Depending
+ *  on the order of optimization steps, we could end up with a scenario as follows:
+ *
+ *      [ProjectionNodeRoot(..., SubqueryExpressionA)]
+ *                                        \
+ *                                         \ references
+ *                                          \
+ *                                        [ProjectionNodeA(..., SubqueryExpressionB)]
+ *                                                                       \
+ *                                                                        \ references
+ *                                                                         \
+ *                                                                        [ProjectionNodeB(...)]
+ *
+ *        (1) collect_subquery_expressions_by_lqp(ProjectionNodeRoot)
+ *                =>  returns { [ ProjectionNodeA, { SubqueryExpressionA } ],
+ *                              [ ProjectionNodeB, { SubqueryExpressionB } ] }
+ *
+ *        (2) OptimizerRuleXY is applied to ProjectionNodeA
+ *                => As a result, SubqueryExpressionB gets replaced / removed from ProjectionNodeA.
+ *
+ *        (3) OptimizerRuleXY is applied to ProjectionNodeB
+ *            -> Wasted optimization time because SubqueryExpressionB and ProjectionNodeB are no longer being used,
+ *               thanks to step (2).
+ *
+ *  With weak pointers, we are forced to skip step (3) because SubqueryExpressionB and its
+ *  corresponding LQP have already been deleted from memory after step (2).
```

Comment:
The last review comment was not about "removed" vs "deleted", but about "memory". Just remove "from memory"?